src/main/java/org/folio/service/impl/StagingUserService.java

```diff
@@ -65,26 +67,36 @@ public Future<User> mergeOrCreateUserFromStagingUser(String stagingUserId, Strin
-        .compose(compositeFuture -> {
-          AddressType homeAddress = compositeFuture.resultAt(0);
-          String homeAddressId = homeAddress.getId();
-
-          Usergroup remotePatronGroup = compositeFuture.resultAt(1);
-
-          return fetchStagingUserById(conn, stagingUserId)
-            .compose(stagingUser -> userId != null ? updateExistingUserDetailsFromStagingUser(stagingUser, userId, conn, homeAddressId)
-              : createNewUserFromStagingUser(stagingUser, conn, homeAddressId, remotePatronGroup))
-            .compose(userUpdatesStorage -> deleteStagingUser(conn, stagingUserId, userUpdatesStorage));
-        })).onSuccess(userUpdatesStorage-> {
-          if(userId == null) {
-            userEventPublisher(vertxContext, okapiHeaders)
-                    .publishCreated(userUpdatesStorage.getNewEntity().getId(), userUpdatesStorage.getNewEntity());
-          } else {
-            userEventPublisher(vertxContext, okapiHeaders)
-                    .publishUpdated(userId, userUpdatesStorage.getOldEntity(), userUpdatesStorage.getNewEntity());
-          }
-        }).compose(userUpdatesStorage-> Future.succeededFuture(userUpdatesStorage.getNewEntity()))
+      .withTrans(conn -> fetchStagingUserById(conn, stagingUserId)
+        .compose(stagingUser -> {
+
+          List<Future<?>> requiredFetches = List.of(
+            fetchHomeAddressType(conn),
+            fetchRemotePatronGroupOnlyIfNotMinorAndUserIdNotPresent(conn, userId, stagingUser.getMinor()),
+            fetchBasicMinorPatronGroupOnlyIfMinorTrue(conn, stagingUser.getMinor())
+          );
+
+          return Future.all(requiredFetches).compose(compositeFuture -> {
+            AddressType homeAddress = compositeFuture.resultAt(0);
+            String homeAddressId = homeAddress.getId();
+            Usergroup remotePatronGroup = compositeFuture.resultAt(1);
+            Usergroup basicMinorPatronGroup = compositeFuture.resultAt(2);
+
+            Future<StagingUserUpdatesStorage<User>> createOrUpdateUser = (userId != null ?
+              updateExistingUserDetailsFromStagingUser(stagingUser, userId, conn, homeAddressId, basicMinorPatronGroup)
+              : createNewUserFromStagingUser(stagingUser, conn, homeAddressId, remotePatronGroup, basicMinorPatronGroup));
+            return createOrUpdateUser.compose(userUpdatesStorage -> deleteStagingUser(conn, stagingUserId,
+              userUpdatesStorage));
+          });
+        })).onSuccess(userUpdatesStorage -> {
+        if (userId == null) {
+          userEventPublisher(vertxContext, okapiHeaders)
+            .publishCreated(userUpdatesStorage.getNewEntity().getId(), userUpdatesStorage.getNewEntity());
+        } else {
+          userEventPublisher(vertxContext, okapiHeaders)
+            .publishUpdated(userId, userUpdatesStorage.getOldEntity(), userUpdatesStorage.getNewEntity());
+        }
```

Comment:
Please consider extracting this code block to a dedicated method, `publishUserEvent` or something called like this